local-volume/provisioner/pkg/common/common.go

```diff
@@ -167,6 +165,7 @@ func CreateLocalPVSpec(config *LocalPVConfig) *v1.PersistentVolume {
 				v1.ReadWriteOnce,
 			},
 			StorageClassName: config.StorageClass,
+			VolumeMode:       &config.VolumeMode,
```

Comment:
So one annoying issue is that the API server will silently drop volume mode if the feature is not enabled.  So this PV will be created with mode filesystem, which could cause problems when we try to cleanup.

I wonder actually if the mount will succeed.  It just does a bind mount of the path.  Could you try that out?

Reply:
The symbolic link in the discovery directory is to the block device, so a bind mount would fail. I did a test by not enabling block feature gate. The PV got created with volume mode missing as you said. When the Pod is started, the kubelet generates "not a directory" error messages:

```
kubernetes-minion-group-018m kubelet[1539]: E0220 23:33:25.288952    1539 nestedpendingoperations.go:267] Operation for "\"kubernetes.io/local-volume/6d5951ac-1696-11e8-8a99-42010a8a0002-local-pv-404316d9\" (\"6d5951ac-1696-11e8-8a99-42010a8a0002\")" failed. No retries permitted until 2018-02-20 23:33:33.288898195 +0000 UTC m=+5850.778604497 (durationBeforeRetry 8s). Error: "MountVolume.SetUp failed for volume \"local-pv-404316d9\" (UniqueName: \"kubernetes.io/local-volume/6d5951ac-1696-11e8-8a99-42010a8a0002-local-pv-404316d9\") pod \"fstest\" (UID: \"6d5951ac-1696-11e8-8a99-42010a8a0002\") : mount failed: exit status 32\nMounting command: systemd-run\nMounting arguments: --description=Kubernetes transient mount for /var/lib/kubelet/pods/6d5951ac-1696-11e8-8a99-42010a8a0002/volumes/kubernetes.io~local-volume/local-pv-404316d9 --scope -- mount -o bind /mnt/disks/ssd0 /var/lib/kubelet/pods/6d5951ac-1696-11e8-8a99-42010a8a0002/volumes/kubernetes.io~local-volume/local-pv-404316d9\nOutput: Running scope as unit: run-r775f02cce68b49f5a2eb14fe75dec352.scope\nmount: mount(2) failed: /var/lib/kubelet/pods/6d5951ac-1696-11e8-8a99-42010a8a0002/volumes/kubernetes.io~local-volume/local-pv-404316d9: Not a directory\n\n"

kubernetes-minion-group-018m kubelet[1539]: mount: mount(2) failed: /var/lib/kubelet/pods/6d5951ac-1696-11e8-8a99-42010a8a0002/volumes/kubernetes.io~local-volume/local-pv-404316d9: Not a directory
```